### PR TITLE
MEJO-19 Implement a dynamically growing tree data structure

### PR DESCRIPTION
**Pull request by pVer297**

### CODE REVIEW

File: Nucleo_F746ZG_app/Src/dynamic_tree.c [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+tree_struct_t create_tree()
+{
+  tree_node_t* root = (tree_node_t*) pvPortMalloc(sizeof(tree_node_t));
+  set_child_null(root);
+  root->parent = NULL;
+  root->num_of_children = 0;
+  root->is_discovered = DISCOVER_TRUE;
+
+  tree_struct_t tree;
+
```
Comment (Rozsnya) [inline Nucleo_F746ZG_app/Src/dynamic_tree.c]: én ezt a sort kiszedném

File: Nucleo_F746ZG_app/Src/dynamic_tree.c [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  tree.root = root;
+  tree.selected_node = root;
+  return tree;
+}
+
+void add_crossroad(tree_struct_t* tree, uint8_t num_of_paths)
+{
+  if ((num_of_paths < 1) || (num_of_paths > 3) ||
+      (tree->selected_node->num_of_children != 0)) {
```
Comment (Rozsnya) [inline Nucleo_F746ZG_app/Src/dynamic_tree.c]: én ezt a sort hozzáraknám az előzőhöz

File: Nucleo_F746ZG_app/Inc/dynamic_tree.h [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  RIGHT_CHILD_NODE
+} tree_node_side_t;
+
+typedef enum {
+  DISCOVER_FALSE,
+  DISCOVER_TRUE
+} tree_boolean_t;
+
+typedef struct tree_node {
+  struct tree_node* child[3];
```
Comment (johnyka) [inline Nucleo_F746ZG_app/Inc/dynamic_tree.h]: variable name should be children

File: Nucleo_F746ZG_app/Inc/dynamic_tree.h [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+typedef enum {
+  LEFT_CHILD_NODE,
+  CENTER_CHILD_NODE,
+  RIGHT_CHILD_NODE
+} tree_node_side_t;
+
+typedef enum {
+  DISCOVER_FALSE,
+  DISCOVER_TRUE
+} tree_boolean_t;
```
Comment (johnyka) [inline Nucleo_F746ZG_app/Inc/dynamic_tree.h]: just use uint8_t as you do in tree_node structure

File: Nucleo_F746ZG_app/Src/dynamic_tree.c [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+void choose_path(tree_struct_t* tree, tree_node_side_t side)
+{
+  uint8_t nodes = tree->selected_node->num_of_children;
+  if ((nodes == 1 && side == LEFT_CHILD_NODE) ||
+      (nodes == 2 && side <= CENTER_CHILD_NODE) ||
+      (nodes == 3 && side <= RIGHT_CHILD_NODE)) {
+    tree->selected_node = tree->selected_node->child[side];
+  }
```
Comment (johnyka) [inline Nucleo_F746ZG_app/Src/dynamic_tree.c]: in situations like this pls provide some warning as a log message

File: Nucleo_F746ZG_app/test/test_dynamic_tree.c [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+  step_back(&test_tree);
+
+  choose_path(&test_tree, RIGHT_CHILD_NODE);
+  discover(&test_tree);
+  step_back(&test_tree);
+
+  choose_path(&test_tree, LEFT_CHILD_NODE);
+  add_crossroad(&test_tree, 1);
+  choose_path(&test_tree, LEFT_CHILD_NODE);
+  //discover(&test_tree);
```
Comment (johnyka) [inline Nucleo_F746ZG_app/test/test_dynamic_tree.c]: pls delete comments

**Pull request merged**